File: 08-DesignPatterns-Java/Solution/Decorator/src/Decorators/LoggingDecorator.java
package Decorators;

import Calculator.ICalculator;

public final class LoggingDecorator implements ICalculator {
    private final ICalculator _decoratedCalculator;

    public LoggingDecorator(ICalculator decoratedCalculator) {
        _decoratedCalculator = decoratedCalculator;
    }

    @Override
    public void calculate() {
        System.out.println(getClass().getName() + ": Before calling the decorated object");
        _decoratedCalculator.calculate();
        System.out.println(getClass().getName() + ": After calling the decorated object");
    }
}

File: 10-DesignPatterns-CSharp/10-2-3-Decorator-Solution/Decorators/LoggingDecorator.cs
namespace Decorator.Decorators
{
    using System;

    public sealed class LoggingDecorator : ICalculator
    {
        private readonly ICalculator _decoratedCalculator;

        public LoggingDecorator(ICalculator decoratedCalculator)
        {
            _decoratedCalculator = decoratedCalculator;
        }

        public void Calculate()
        {
            Console.WriteLine($"{nameof(LoggingDecorator)}: Before calling the decorated object");
            _decoratedCalculator.Calculate();
            Console.WriteLine($"{nameof(LoggingDecorator)}: After calling the decorated object");
        }
    }
}
